compiler/hash-attrs/src/target.rs

```diff
@@ -0,0 +1,292 @@
+        const Lit = 1 << 5;
+
+        /// A loop expression, representing `loop`, `while` and `for` expressions.
+        const Loop = 1 << 6;
+
+        /// A match block, represents for `match` and `if` expressions.
+        const Match = 1 << 7;
+
+        /// An implementation definition block.
+        const ImplDef = 1 << 8;
+
+        /// A top-level module.
+        const Mod = 1 << 9;
+
+        /// A module definition block.
+        const ModDef = 1 << 10;
+
+        /// A block, specifically the [`hash_ast::ast::Block::Body`] variant.
+        const Block = 1 << 11;
+
+        /// An `import` statement.
+        const Import = 1 << 12;
+
+        /// A type function definition.
+        const TyFnDef = 1 << 13;
+
+        /// A `struct` definition.
+        const StructDef = 1 << 14;
+
+        /// An `enum` definition.
+        const EnumDef = 1 << 15;
+
+        /// An enum variant.
+        const EnumVariant = 1 << 16;
+
+        /// A function definition, regardless of the position.
+        const FnDef = 1 << 17;
+
+        /// A type.
+        const Ty = 1 << 18;
+
+        /// A type argument, whilst this has similar implications to
+        /// a type, it adds additional context that this is a type
+        /// argument.
+        const TyArg = 1 << 19;
+
+        /// A parameter in a struct, enum, tuple or function definition.
+        const Param = 1 << 20;
+
+        /// A field in a struct or enum.
+        const Field = 1 << 21;
+
+        /// A match branch.
+        const MatchCase = 1 << 22;
+
+        /// A pattern.
+        const Pat  = 1 << 23;
+
+        /// A pattern argument.
+        const PatArg  = 1 << 24;
+
+        /// A general item definition e.g. `struct`, `enum`, `impl`, `mod` and `fn`.
+        const Item = Self::StructDef.bits() | Self::EnumDef.bits() | Self::FnDef.bits() | Self::TyFnDef.bits() | Self::ImplDef.bits() | Self::ModDef.bits();
+    }
+}
+
+impl AttrTarget {
+    /// Classify the given [ast::Expr] into a [AttrTarget].
+    pub fn classify_expr(expr: &ast::Expr) -> Self {
+        match expr {
+            ast::Expr::ConstructorCall(_) => AttrTarget::ConstructorCall,
+            ast::Expr::Macro(_) => AttrTarget::MacroInvocation,
+            ast::Expr::Unsafe(_) => AttrTarget::Unsafe,
+            ast::Expr::Lit(_) => AttrTarget::Lit,
+            ast::Expr::Block(ast::BlockExpr { data: block }) => match block.body() {
+                ast::Block::Loop(_) | ast::Block::While(_) | ast::Block::For(_) => AttrTarget::Loop,
+                ast::Block::Match(_) | ast::Block::If(_) => AttrTarget::Match,
+                ast::Block::Body(_) => AttrTarget::Block,
+            },
+            ast::Expr::Import(_) => AttrTarget::Import,
+            ast::Expr::StructDef(_) => AttrTarget::StructDef,
+            ast::Expr::EnumDef(_) => AttrTarget::EnumDef,
+            ast::Expr::TyFnDef(_) => AttrTarget::TyFnDef,
+            ast::Expr::TraitDef(_) => AttrTarget::Item,
```

Comment:
Should other defs also be items?

Reply:
This is a mistake, it should be just trait def.